# Chapter Three: Julia Basics

Note that I am skipping Chapters One and Two because there isn't much code in those chapters. I suggest reading them for completeness though. 

## A Note on Projects

If you are like me and want to work with a Julia project, start it up. See the `Readme.md` file in this repo for how to create a project. We will be adding to this project as we go through these notebooks and all add-on packages will be stored here. Within each notebook, you have to load `Pkg` and activate the specific project that we are using. This keeps track of all of our installed packages for this project. This first chapter really only uses standard packages (I think), so it's not terribly important. I'll keep it here nevertheless. Here you go. 

In [1]:
using Pkg
Pkg.activate("juliadatascience")

  Activating project at `~/research/JuliaDataScience/notebooks/juliadatascience`


## 3.1 Development Environments

In [2]:
x = 2

2

In [3]:
x + 1

3

We created a file named `script.jl` with the following code. 

```
x = 3
y = 4
```

We can use the `include` statement to execute the code in this file, i.e. load the file into Julia. 

In [4]:
include("script.jl")

4

In [5]:
x

3

In [6]:
y

4

## 3.2 Language Syntax

### 3.2.1 Variables

In [7]:
name = "Julia"
age = 9

9

In [8]:
name

"Julia"

In [9]:
age = 10

10

In [10]:
name = "Julia Rapidus"

"Julia Rapidus"

In [11]:
12 * age

120

In [12]:
typeof(age)

Int64

Note that by default both integers and real numbers are stored using 64 bits. You can define more or less bits if needed. 

This next command is super useful. I always forget in python what functions or methods can actually be applied to variables objects. The authors of the book describe this operation for integers as “What else can I do with integers?” In general, if you have an object, just replace whatever object with integer in the next code chunk. The `first` function will restrict the output to the first five elements in the vector.

In [13]:
first(methodswith(Int64), 5)

[1] AbstractFloat(x::Int64) in Base at float.jl:243
[2] Float16(x::Int64) in Base at float.jl:146
[3] Float32(x::Int64) in Base at float.jl:146
[4] Float64(x::Int64) in Base at float.jl:146
[5] Int64(x::Union{Bool, Int32, Int64, UInt32, UInt64, UInt8, Int128, Int16, Int8, UInt128, UInt16}) in Core at boot.jl:764

### 3.2.2 User-Defined Types

In Julia, you can create your own user-defined data type called a `struct`. Here's the first one that the book defines. It's a data object for storing information about programming languages in fields. To view the fields of a `struct`, you use the `fieldnames(your_named_struct)` command.  

In [14]:
struct Language
    name::String
    title::String
    year_of_birth::Int64
    fast::Bool
end

In [15]:
fieldnames(Language)

(:name, :title, :year_of_birth, :fast)

The `struct` is just an object type until you define them using the `=` sign. For example, here are the two `struct`s named ulia and python...clever.

In [16]:
julia = Language("Julia", "Rapidus", 2012, true)
python = Language("Python", "Letargicus", 1991, false)

Language("Python", "Letargicus", 1991, false)

(Aside: One thing I don't necessarily like about notebooks is how it returns stuff (output) from the last line in a cell. Why? What's the point?)

Good luck changing the values of those objects we just defined! You can't. Those things are fixed in place once defined and if you need a `struct` that can be changed, use a `mutable struct` instead, e.g. here's `MutableLanguage`. 

In [17]:
mutable struct MutableLanguage
    name::String
    title::String
    year_of_birth::Int64
    fast::Bool
end

julia_mutable = MutableLanguage("Julia", "Rapidus", 2012, true)

MutableLanguage("Julia", "Rapidus", 2012, true)

Now, if we want to change a field in the `mutable struct`, e.g. the `title` field, we just 

In [18]:
julia_mutable.title = "Python Obliteratus"

julia_mutable

MutableLanguage("Julia", "Python Obliteratus", 2012, true)

Boy, they are clever with their python digs. :/ Or they have a serious case of little bro-itis. Whatever. 

### 3.2.3 Boolean Operators and Numeric Comparisons

Boolean operators are pretty standard with the three usuals: not (`!`), and (`&&`), and or (`||`). Boolean's in action:

In [19]:
!false

true

In [20]:
(false && true) || (!false) ## Note the order of operations is obviously standard

true

In [21]:
(6 isa Int64) && (6 isa Real)

true

That last one kind of blows my mind. Julia has the usual comparison operators too, e.g. equality, less than (or equal to), and greater than (or equal to). They are `==`, `<` (or `<=`), and `>` (or `>=`), respectively. You can use `≠`, `≤`, or `≥` signs if you want, but I probably will stick to the `>=` type comparisons. Here are some examples.

In [22]:
1 == 1

true

In [23]:
1 >= 10

false

In [24]:
1 ≥ 10

false

In [25]:
1 == 1.0 # even on different object types

true

In [26]:
typeof(1)

Int64

In [27]:
typeof(1.0)

Float64

What the?

In [28]:
(1 != 10) || (3.14 <= 2.71)

true

### 3.2.4 Functions

Okay, time for some cool stuff. Let's learn about functions. How write/define them, use them, etc. I like their description of functions, "In Julia, a function **maps argument’s values to one or more return values**." Note: You'll get an error if you try to run this code chunk. 

In [29]:
function function_name(arg1, arg2)
    result = stuff with the arg1 and arg2
    return result
end

LoadError: syntax: "function" at In[29]:1 expected "end", got "with"

Note that `arg1` and `arg2` are deleted after the function runs. The only thing that is returned is `return`. Pretty standard stuff. The one-liner version of that function is pretty nice. 

In [30]:
f_name(arg1, arg2) = stuff with the arg1 and arg2

LoadError: syntax: extra token "with" after end of expression

#### 3.2.4.1 Creating new Functions

In [31]:
function add_numbers(x, y)
    return x + y
end

add_numbers (generic function with 1 method)

In [32]:
add_numbers(2, 23)

25

In [33]:
add_numbers(3.14, 2.72)

5.86

You might want to have a function that behaves differently for different types of inputs. For example, suppose a function`round_numbers` should return a rounded value if given floats. You can define multiple functions with the same name, but different inputs. Julia will recognize the appropriate function to invoke depending on its inputs. 

In [34]:
function round_number(x::Float64)
    return round(x)
end

function round_number(x::Int64)
    return x
end

round_number (generic function with 2 methods)

Here's another handy function, `methods`. Again, something that I need a million times when programming in whatever language. Try it out on the `round_number` functions. 

In [35]:
methods(round_number)

# 2 methods for generic function "round_number":
[1] round_number(x::Float64) in Main at In[34]:1
[2] round_number(x::Int64) in Main at In[34]:5

Hmmmm, good point authors of the text. What if we want to add two 8-bit integers or something else? We can create a generic function with an **abstract type** such as `AbstractFloat` or `Integer`. Very handy indeed. 

In [36]:
function round_number(x::AbstractFloat)
    return x
end

round_number (generic function with 3 methods)

In [37]:
methods(round_number)

# 3 methods for generic function "round_number":
[1] round_number(x::Float64) in Main at In[34]:1
[2] round_number(x::AbstractFloat) in Main at In[36]:1
[3] round_number(x::Int64) in Main at In[34]:5

In [38]:
x_32 = Float32(1.1)
round_number(x_32)

1.1f0

More handy features. The `Base.show` function will return a default printing of specific objects, e.g. that `struct` that we defined before for different programming languages. You can specify how you want `Base.show` to operate on specific objects by creating a new `Base.show` for your `struct` if you want. Take a look at the default `Base.show`. 

In [39]:
Base.show(python)

Language("Python", "Letargicus", 1991, false)

In [40]:
show(python)

Language("Python", "Letargicus", 1991, false)

In [41]:
show(julia)

Language("Julia", "Rapidus", 2012, true)

In [42]:
Base.show

show (generic function with 281 methods)

In [43]:
Base.show(io::IO, l::Language) = print(
    io, l.name, ", ",
    2021 - l.year_of_birth, " years old, ",
    "has the following titles: ", l.title
)

In [44]:
Base.show

show (generic function with 282 methods)

In [45]:
Base.show(python)

Python, 30 years old, has the following titles: Letargicus

#### 3.2.4.2 Multiple Return Values

In [46]:
function add_multiply(x, y)
    addition = x + y
    multiplication = x * y
    return addition, multiplication
end

add_multiply (generic function with 1 method)

In [47]:
return_1, return_2 = add_multiply(1, 2)

(3, 2)

In [48]:
return_1

3

In [49]:
return_2

2

I like being able to define multiple items being returned in this way. There is also a `first` and `last` function that we can use from `Base`. 

In [50]:
all_returns = add_multiply(1, 2)

(3, 2)

In [52]:
Base.first(all_returns) ## or just first(all_returns)

3

In [53]:
last(all_returns) ## or Base.last(all_returns)

2

Can we do this with multiple returns? If so, what's the trick? A `tuple` is returned, so you can access the elements with the `[]` notation. Note: Julia uses a 1-based index. 

In [54]:
function add_multiply_divide(x, y)
    addition = x + y
    multiplication = x * y
    division = x/y
    return addition, multiplication, division
end

add_multiply_divide (generic function with 1 method)

In [55]:
all_returns = add_multiply_divide(10,2)
all_returns

(12, 20, 5.0)

In [56]:
first(all_returns)

12

In [59]:
all_returns[2]

20

#### 3.2.4.3 Keyword Arguments

We'll get to keyword arguments in a second. First, how do you know if something is a subtype of another something? Use the `<:` operator. For example, is an `AbstractFloat` a subtype of a `Real`, or `Integer` a subtype of `Real`? Yes. Check it out. 

In [60]:
AbstractFloat <: Real

true

In [61]:
Integer <: Real

true

In [62]:
Real <: Integer

false

OK, here you go for an example about keyword args. It's easy to see in this log function.

In [63]:
function logarithm(x::Real; base::Real=2.7182818284590)
    return log(base, x)
end

logarithm (generic function with 1 method)

In [64]:
logarithm(10; base=2)

3.3219280948873626

In [65]:
logarithm(10; base=10)

1.0

#### 3.2.4.4 Anonymous Functions

Suppose we want to make a quick, throw away-type function. Anonymous functions are the answer. How do we use these things? This function will use `logarithm(2)` as `x` and evaluate `2.7182818284590^logarithm(2)`. 

In [69]:
map(x -> 2.7182818284590^x, logarithm(2))

2.0

In [70]:
2.7182818284590^logarithm(2)

2.0

### 3.2.5 Conditional If-Else-Elseif

Typical behavior for `if`, `else`, and `elseif` conditional evaluations. If it wasn't typical, I'd quit this tutorial now. Here's the example. 

In [75]:
a = 5.0
b = 4

if a < b
    "a is less than b"
elseif a > b
    "a is greater than b"
else
    "a is equal to b"
end

"a is greater than b"

Same thing wrapped in a function, `compare`. 

In [76]:
function compare(a, b)
    if a < b
        "a is less than b"
    elseif a > b
        "a is greater than b"
    else
        "a is equal to b"
    end
end

compare (generic function with 1 method)

In [77]:
compare(5.0, 4)

"a is greater than b"

In [78]:
compare(4, 5)

"a is less than b"

In [79]:
compare(3.14, 3.14)

"a is equal to b"

### 3.2.6 For Loop

Again, for loops are straightforward too. 

In [80]:
for i in 1:10
    println(i)
end

1
2
3
4
5
6
7
8
9
10


### 3.2.7 While Loop

Evaluate the stuff inside the loop while keyword (in this case, `n`) in the conditional statement evaluates to `true`. 

In [83]:
n = 0

while n < 3
    println(n)
    global n += 1
end

0
1
2


In [84]:
n

3

Note the `global` keyword on the `n` within the loop. We add this because variables defined inside a loop (function, conditional expression) only exist within the loop. In this case, we are incrementing the `global n` and need the `while` loop to recognize that fact. 

## 3.3 Native Data Structures

Crap, I forgot several sections from Chapter 3. I already moved on to Chapter 4 by completely neglecting sections 3.3, 3.4, and 3.5. I think that stuff might be useful for the later chapters. This stuff has been mo

## Main Takeaways

OK, what did I learn in this chapter? 
1. The `struct` object seems pretty powerful. Basically building up data objects with differing types of data is useful. Kind of reminds me of a tibble in R if we can have a `struct` of `struct`s. Maybe?  
2. I'll try to work using the [Julia style guide(https://docs.julialang.org/en/v1/manual/style-guide/) as much as possible. However, I really don't like the camel case stuff, e.g. when defining a `struct`, but it is what it is. 
3. I like modifying the `Base.show` function to match the generic `struct` that was defined. 
4. Everything in the latter part of this notebook is pretty straightforward. Good stuff. 